### PR TITLE
Create Helm chart for Kong 

### PR DESCRIPTION
**Pull request by lawrencegripper**
Hi, 

I've done some work to create a [helm](https://docs.helm.sh/) chart to simplify deploying Kong. 

Currently added under a subfolder with some lightweight docs as I wasn't sure how best to document with the exiting yaml files. 

Hopefully useful!

### CODE REVIEW

**Comment by shashiranjan84** [comment]
@lawrencegripper thanks for PR, I will take a look.

**Comment by nrkrishn** [comment]
This is awesome. I would love to see this PR approved and merged. One stumbling block I ran into was that I wasn't able to get the changes in this PR running without creating a job to run the migrations. @lawrencegripper I would be happy to contribute the yaml for that to your fork if you are interested. Thanks.

**Comment by shashiranjan84** [comment]
@nrkrishn yes PR need to be updated for new Kong v0.11, just trying find time to do the change.

**Comment by lawrencegripper** [comment]
@nrkrishn Sure happy to accept the changes to add the migrations job to my fork. Hopefully then gives us a working PR which @shashiranjan84 can review and merge if all looks good. Sound like a good plan?

**Comment by nrkrishn** [comment]
Thanks for the quick turn around guys. @lawrencegripper I forked your fork and created a PR with the yaml file for running migrations. I tested both the postgres and cassandra code paths, but admittedly the cassandra code path took a while and locked up the "kubernetes get all" command for a time being, before eventually bringing everything up in a healthy state.

**Comment by nrkrishn** [comment]
For what it's worth, I believe the issues I mention in my comment above were caused by a lack a system resources on my laptop (I'm testing via minikube). When I ran the helm deploy with cassandra as the backing store it caused the tiller process to crash. As I mention, everything eventually started up healthy. I just had to restart minikube (in order to restart tiller) when I wanted to delete my deployment.

**Comment by shashiranjan84** [comment]
@nrkrishn @lawrencegripper it would need further changes as with Kong v0.11.0 we removed SERF dependency and thus we don't need expose SERF port. Further now Kong can run as Nginx process. You can check the latest manifest files for example.

https://github.com/Mashape/kong-dist-kubernetes/blob/master/kong_postgres.yaml

**Comment by vongohren** [comment]
How is this coming along?

**Comment by nrkrishn** [comment]
@lawrencegripper, I can remove the surf-udp/surf-tcp ports. I can also take a stab at running Kong as an an nginx process, though I'll admit I'm a bit lost... @shashiranjan84, if you don't mind, could you please provide some further clarification/guidance on the nginx part?

**Comment by shashiranjan84** [comment]
@nrkrishn You can use the https://github.com/Mashape/kong-dist-kubernetes/blob/master/kong_postgres.yaml manifest file as reference, If you there is no explicit command section. Official images with take care of how the Kong process is started.

Also I noticed 3rd party dashboard is included in chart. It need to go as we do not support it officially. 
https://github.com/lawrencegripper/kong-dist-kubernetes/blob/76af484689edfea954ae6c086a6299bf79f2988c/charts/kong/templates/NOTES.txt#L20

**Comment by lawrencegripper** [comment]
Cool, sorry I missed your PR @nrkrishn which adds migrations. I've now merged that in to my fork now.

@shashiranjan84 I find the dash useful but understand support requirements. Would it be acceptable to make it an optional feature, off by default and add some text explaining it's unsupported? If not I'll remove and push. 

Will check for any other changes between yaml's and then hopefully good to go. Little busy my side so may be a few days, hope thats alright?

**Comment by shashiranjan84** [comment]
@lawrencegripper sorry, it can't be optional too.We always appreciate your contribution, thanks.

**Comment by ipedrazas** [comment]
@lawrencegripper Have you thought about submitting this chart to the helm charts repo? https://github.com/kubernetes/charts it would give greater visibility 😄 

If you send a PR to _incubator_ it will be merged so people can test and use it. Once it's in incubator we can iterate on the best practices and have a great Kong chart in _stable_.

File: charts/kong/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        app:  {{ .Release.Name }}-{{ .Values.kongName }}
+    spec:
+      containers:
+      - name: kong
+        image: kong
+        env:
+          - name: KONG_DATABASE
+            value: {{ .Values.backingStoreConfig.KONG_DATABASE}}
+          - name: KONG_LOG_LEVEL
+            value: debug
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply charts/kong/templates/deployment.yaml]: Agreed I'll make the log_level configurable

File: charts/kong/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            value: {{ .Release.Name }}-{{ .Values.backingStoreConfig.KONG_PG_HOST }}.{{ .Release.Namespace }}.svc.cluster.local
+            {{ else }}
+          - name: KONG_PG_HOST
+            value: {{ .Values.backingStoreConfig.KONG_PG_HOST }}
+            {{ end }}
+        {{ end }}
+        command: [ "/bin/sh", "-c", "KONG_CLUSTER_ADVERTISE=$(KONG_HOST_IP):7946 KONG_NGINX_DAEMON='off' kong start --vv" ]
+        ports:
+        - name: admin
+          containerPort: 8001
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply charts/kong/templates/deployment.yaml]: The ports are configurable via the services.yaml and values.yaml - think this is better as other containers will connect via these

File: charts/kong/templates/kongdashboard.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+{{ if .Values.kongDashboard.deploy }}
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply charts/kong/templates/kongdashboard.yaml]: Will remove

File: charts/kong/templates/postgres.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
+{{ if and (.Values.backingStoreConfig.deployment.deployStore) (eq .Values.backingStoreConfig.KONG_DATABASE "postgres") }}
+
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ .Release.Name }}-{{ .Values.backingStoreConfig.KONG_PG_HOST }}
+spec:
+  ports:
+  - name: pgql
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply charts/kong/templates/postgres.yaml]: Added config for KONG_PG_PORT - is this the correct env variable?

File: charts/kong/templates/postgres.yaml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    port: 5432
+    targetPort: 5432
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply charts/kong/templates/postgres.yaml]: Added config for KONG_PG_POR

File: charts/kong/templates/postgres.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply charts/kong/templates/postgres.yaml]: Sorry, I'm unclear what you'd like to be configurable here.

File: charts/kong/templates/service.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
+apiVersion: v1
```
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply charts/kong/templates/service.yaml]: Sure will do.

**Comment by lawrencegripper** [comment]
@ipedrazas I think this is a good idea. I'll take this as an action to do now. 

@magicalbanana I've updated the code based on your comments and pushed now. Need to run some more testing this afternoon but wanted to share now given my delay up to this point.

File: charts/kong/values.yaml [hunk 7]
```diff
@@ -36,7 +36,7 @@ backingStoreConfig:
   KONG_PG_PASSWORD: kong
   KONG_PG_DATABASE: kongdb
   KONG_PG_HOST: kong-postgresql
-  KONG_LOG_LEVEL: 
+  KONG_LOG_LEVEL: debug #Options: debug, info, notice, warn, error, crit, alert, emerg
```
Comment (magicalbanana) [inline charts/kong/values.yaml]: Cool.

File: charts/kong/templates/postgres.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (magicalbanana) [reply charts/kong/templates/postgres.yaml]: The name of the port, perhaps you can make this configurable.

File: charts/kong/templates/deployment.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (magicalbanana) [reply charts/kong/templates/deployment.yaml]: Hmm, that's true. But in the event that someone wants to configure the ports it connects to, it's also good to have them defaulted. Don't you think so?

File: charts/kong/templates/postgres.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (lawrencegripper) [reply charts/kong/templates/postgres.yaml]: I'm not sure this is necessary as I don't think you would ever need to customize it. The stable postgres chart, for example, doesn't have this configurable [Postgres Chart](https://github.com/kubernetes/charts/blob/master/stable/postgresql/templates/deployment.yaml#L48)

**Comment by lawrencegripper** [comment]
Hopefully now all set, are we alright to get this PR merged?

**Comment by lawrencegripper** [comment]
Nudge

File: charts/kong/README.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+3. edit the "values.yaml" doc to fit your environment
+3. helm package .
+4. helm install kong-0.1.1.tgz --name yourreleasename
+
+You will then see notes on how to connect and use your Kong deployment. 
+
+Todo:
+--------
+
+- Fix cassandra deployment issues
```
Comment (shashiranjan84) [inline charts/kong/README.md]: @lawrencegripper What wrong with Cassandra deployment?

**Comment by shashiranjan84** [comment]
@lawrencegripper would you please squash all commits into one.

**Comment by shashiranjan84** [comment]
@lawrencegripper I created a branch based of your branch, I will make some minor changes before merging it.

File: charts/kong/README.md (hunk 8 above)
Comment (ipedrazas) [reply charts/kong/README.md]: @shashiranjan84 how many times have you deployed Cassandra in Kubernetes?

Comment (shashiranjan84) [reply charts/kong/README.md]: @ipedrazas none with this chart, currently I am in process of updating this PR, please check https://github.com/Kong/kong-dist-kubernetes/tree/charts. I will make sure Kong works with both Postgres and Cassandra.

**Comment by stackedsax** [comment]
Thanks for putting this together, @lawrencegripper and @shashiranjan84.

A couple quick questions: should this chart use a StatefulSet rather than a ReplicationController for Cassandra in this day and age?  It seems like that's where recent Cassandra examples and helm charts have gone.  Perhaps just put the cassandra chart from the chart repo that @ipedrazas mentioned earlier in a requirements.yaml and use that directly?

Thanks!

**Comment by shashiranjan84** [comment]
@stackedsax I am in process of updating this PR,  change already been made to use statefulset https://github.com/Kong/kong-dist-kubernetes/blob/charts/charts/kong/templates/cassandra.yaml. Still need to work documentation and need some testing which I will try to finish once I have some time.

**Comment by stackedsax** [comment]
Thanks for pointing out your branch and thanks for updating this to a StatefulSet.

Any reason not to just use https://github.com/kubernetes/charts/blob/master/incubator/cassandra/templates/cassandra-statefulset.yaml? Seems like it has a whole bunch of little goodies that might be nice to use.

I'll test our your branch in the morning.  Thanks!

**Comment by shashiranjan84** [comment]
Just not getting time, I would love to give it a try and include in PR.

**Comment by shashiranjan84** [comment]
@lawrencegripper @stackedsax finally I got some time to work/learn on Helm. It made more sense 
 to pull the official chart for `Cassandra` and `Postgresql`. 

Now I would need some help from you guys to test the branch https://github.com/Kong/kong-dist-kubernetes/tree/charts.

**Comment by stackedsax** [comment]
Thanks for the progress!  I tested your branch a couple days ago and it worked reasonably well at that point.  

I was going to mention again the idea of pulling things in via requirements.yaml, so I'm glad you've gone that way.  Those charts in the helm repo just had some nice bells and whistles that you were going to have to re-do if you wanted them, too.

I'll try and find time tomorrow to test the new stuff.

**Comment by shashiranjan84** [comment]
Closing this PR in favor of https://github.com/Kong/kong-dist-kubernetes/pull/29.

**Pull request closed without merging**